source/extensions/common/tap/tap_matcher.h

```diff
@@ -298,6 +343,92 @@ class HttpResponseTrailersMatcher : public HttpHeaderMatcherBase {
   }
 };
 
+/**
+ * Base class for body matchers.
+ */
+class HttpBodyMatcherBase : public SimpleMatcher {
+public:
+  HttpBodyMatcherBase(const std::vector<MatcherPtr>& matchers) : SimpleMatcher(matchers) {}
+
+protected:
+  // Limit search to specified number of bytes.
+  // Value equal to zero means no limit.
+  uint32_t limit_{};
+};
+
+/**
+ * Context is used by HttpGenericBodyMatcher to:
+ * - track how many bytes has been processed
+ * - track patterns which have been found
+ * - store last several seen bytes of the HTTP body (when pattern starts at the end of previous body
+ *   chunk and continues at the beginning of the next body chunk)
+ */
+class HttpGenericBodyMatcherCtx : public MatcherCtx {
+public:
+  HttpGenericBodyMatcherCtx(std::list<std::string> patterns, size_t overlap_size)
+      : patterns_(patterns) {
+    // Initialize overlap_ buffer's capacity to fit the longest pattern - 1.
+    overlap_.reserve(overlap_size);
+  }
+  ~HttpGenericBodyMatcherCtx() override = default;
+
+  // List is initialized based on the Matcher's configuration.
+  // When data is passing through the matcher, found patterns are removed
+  // from the list. When all patterns have been found, the list is empty.
+  std::list<std::string> patterns_;
```

Comment:
So this is copying each pattern for each request, and (if it does not match) storing it for the lifetime of the stream?  IF we're going to go this route I think we probably ought to configure-enforce limits on the bytes allowed to check, because this could get really memory expensive really fast.
Alternately, as much as I dislike shared pointers, it might make sense to stick matcher config in a shared pointer and track references rather than do full copies.

Reply:
Good point. For some reason I imagined that patterns will not be that large. I will see if a shared pointer can be used here.  Maybe instead of copying patterns, the matcher will store indexes of patterns not found yet. It will be nice reduction of memory and there will be only one copy of patterns shared across all instances of matchers. Thanks for pointing this out.

Reply:
Changed the code as described above: individual instances of per-stream matchers store shared pointer to a set of patterns plus a list of indexes to entries in that shared pointer. When a pattern has been found, index is removed. This solves the problem with memory and each pattern exists only once in memory regardless of the number of per-stream matchers.